File: Catter2-Dagger2Sample/app/src/main/java/io/catter2/di/AppComponent.java
package io.catter2.di;

import android.content.Context;

import javax.inject.Singleton;

import dagger.Component;
import io.catter2.cat_api.TheCatAPI;

/**
 * Created by robert on 20.9.2017.
 */
@Singleton
@Component(modules = {AppModule.class, TheCatAPIModule.class})
public abstract class AppComponent {
    private static AppComponent instance;

    public static AppComponent get() {
        return AppComponent.instance;
    }

    public static void initialize(AppModule appModule, TheCatAPIModule theCatAPIModule) {
        if (AppComponent.get() != null) {
            throw new RuntimeException("AppComponent already initialized");
        }

        AppComponent.instance = DaggerAppComponent.builder()
                .appModule(appModule)
                .theCatAPIModule(theCatAPIModule)
                .build();
    }

    abstract public Context getAppContext();

    abstract public TheCatAPI getTheCatAPI();
}

File: MvvmSampleApp/app/src/main/java/com/example/robert/mvvmsampleapp/di/AppComponent.kt
package com.example.robert.mvvmsampleapp.di

import android.app.Application
import com.example.robert.mvvmsampleapp.MVVMApplication
import dagger.BindsInstance
import dagger.Component
import dagger.android.AndroidInjectionModule
import javax.inject.Singleton

/**
 * Created by robert on 23.9.2017.
 */
@Singleton
@Component(modules = arrayOf(AndroidInjectionModule::class, AppModule::class, MainActivityModule::class))
interface AppComponent {
    @Component.Builder
    interface Builder {
        @BindsInstance
        fun application(application: Application): Builder

        fun build(): AppComponent
    }

    fun inject(application: MVVMApplication)
}